pkg/driver/node.go

```diff
@@ -0,0 +1,566 @@
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// FSTypeExt2 represents the ext2 filesystem type
+	FSTypeExt2 = "ext2"
+	// FSTypeExt3 represents the ext3 filesystem type
+	FSTypeExt3 = "ext3"
+	// FSTypeExt4 represents the ext4 filesystem type
+	FSTypeExt4 = "ext4"
+	// FSTypeXfs represents te xfs filesystem type
+	FSTypeXfs = "xfs"
+
+	defaultFsType = FSTypeExt4
+
+	MaxRetryCount = 10
+)
+
+var (
+	ValidFSTypes = []string{FSTypeExt2, FSTypeExt3, FSTypeExt4, FSTypeXfs}
+)
+
+var (
+	// nodeCaps represents the capability of node service.
+	nodeCaps = []csi.NodeServiceCapability_RPC_Type{
+		csi.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME,
+	}
+)
+
+// node is the server implementation
+// for CSI NodeServer
+type node struct {
+	client           *client.Client
+	driver           *CSIDriver
+	mounter          *NodeMounter
+	volumeTransition *request.Transition
+}
+
+// NewNode returns a new instance
+// of CSI NodeServer
+func NewNode(d *CSIDriver, cli *client.Client) csi.NodeServer {
+	return &node{
+		client:           cli,
+		driver:           d,
+		mounter:          newNodeMounter(),
+		volumeTransition: request.NewTransition(),
+	}
+}
+
+func (ns *node) attachDisk(instance *jv.JivaVolume) (string, error) {
+	connector := iscsi.Connector{
+		VolumeName:    instance.Name,
+		TargetIqn:     instance.Spec.ISCSISpec.Iqn,
+		Lun:           instance.Spec.ISCSISpec.Lun,
+		Interface:     instance.Spec.ISCSISpec.ISCSIInterface,
+		TargetPortals: instance.Spec.ISCSISpec.TargetPortals,
+		Discovery:     true,
+	}
+
+	logrus.Debugf("NodeStageVolume: attach disk with config: {%+v}", connector)
+	devicePath, err := iscsi.Connect(connector)
+	if err != nil {
+		return "", err
+	}
+
+	if devicePath == "" {
+		return "", fmt.Errorf("connect reported success, but no path returned")
+	}
+	return devicePath, err
+}
+
+func (ns *node) waitForVolumeToBeReady(volID string) (*jv.JivaVolume, error) {
+	var retry int32
+	for {
+		if err := ns.client.Set(); err != nil {
+			return nil, err
+		}
+
+		instance, err := ns.client.GetJivaVolume(volID)
+		if err != nil {
+			return nil, err
+		}
+
+		time.Sleep(2 * time.Second)
+		retry++
+		if instance.Status.Phase == jv.JivaVolumePhaseCreated {
+			return instance, nil
+		} else if retry > MaxRetryCount {
+			if instance.Status.Status == "RO" {
+				status := instance.Status.ReplicaStatuses
+				if status != nil {
+					return nil, fmt.Errorf("Volume is RO: replica status: %+v", status)
+				}
+				return nil, fmt.Errorf("Volume is not ready: replicas may not be connected")
+			}
+			return nil, fmt.Errorf("Volume is not ready: volume status is %s", instance.Status.Status)
+		}
+	}
+}
+
+func (ns *node) waitForVolumeToBeReachable(targetIP string) error {
+	var (
+		retries int
+		err     error
+		conn    net.Conn
+	)
+
+	for {
+		// Create a connection to test if the iSCSI Portal is reachable,
+		if conn, err = net.Dial("tcp", targetIP); err == nil {
```

Comment:
Lets rename the above variable to targetIPPort or targetPortal. targetIP looks to be misleading. It looks as if we are creating a TCP connection without providing any targetPort.

Reply:
sounds good